File: November_Leetcode_Challenge_2022/Day2/433._Minimum_Genetic_Mutation.java
class Solution {
    public int minMutation(String start, String end, String[] bank) {
        char [] c= new char[]{'A','C','G','T'};
        Queue<String> que = new LinkedList<>();
        Set<String> visited= new HashSet<>();
        int count=0;
        que.add(start);
        visited.add(start);
        while(!que.isEmpty()){
            int size=que.size();
            while(size-->0){
                String temp=que.poll();
                if(temp.equals(end)){
                    return count;
                }
                for(char tempc: c){
                    for(int i=0;i<8;i++){
                         String tempstring = temp.substring(0, i) + tempc + temp.substring(i + 1);
                        if(!visited.contains(tempstring) && Arrays.asList(bank).contains(tempstring)){
                            que.add(tempstring);
                            visited.add(tempstring);
                        }
                    }
                }
            }
            count++;
        }
        return -1;
    }
    
}

File: November_Leetcode_Challenge_2022/Day2/433._Minimum_Genetic_Mutation.cpp
class Solution {
public:
    int minMutation(string start, string end, vector<string>& bank) {
        queue<string>q;
        unordered_set<string>visited;
        
        q.push(start);
        visited.insert(start);
        
        int count = 0;
        
        while(!q.empty())
        {
            int n = q.size();
            for(int i=0;i<n;i++)
            {
                string node = q.front();
                q.pop();
                
                if(node == end)
                {
                    return count;
                }
                
                for(char ch :"ACGT")
                {
                    for(int j=0;j<node.size();j++)
                    {
                        string adjNode = node;
                        adjNode[j] = ch;
                       if(!visited.count(adjNode) &&                  find(bank.begin(),bank.end(),adjNode)!= bank.end())
            {
                q.push(adjNode);
                visited.insert(adjNode);
            }
                    }
                }
            }
             count++;
        }
       return -1; 
    }
};
